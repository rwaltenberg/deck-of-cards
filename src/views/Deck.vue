<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Deck from '@/utils/deck'
import Desk from '@/components/Desk.vue'

@Component({
  components: {
    Desk
  }
})
export default class DeckDetails extends Vue {
  private deck: Deck | null = null

  get sorted () {
    return this.deck && {
      cards: this.deck.getSortedCards(),
      string: this.deck.getSortedString()
    }
  }

  get fullHouses () {
    return this.deck && this.deck.getFullHouseStrings()
  }

  async mounted () {
    try {
      this.deck = await Deck.fromDeckId(this.$route.params.id)
    } catch (error) {
      console.error(error)
    }
  }
}
</script>

<template>
  <div v-if="deck" class="deck">
    <desk :value="sorted.cards" />

    <dl class="stats">
      <dt>High Card:</dt>
      <dd>{{ sorted.string[0] }}</dd>
      <dt>Full House Combo:</dt>
      <dd :class="{ 'has-full-houses': fullHouses.length }">
        <template v-if="!fullHouses.length">None</template>
        <ol class="full-houses" v-else>
          <li v-for="set in fullHouses" :key="set">{{ set }}</li>
        </ol>
      </dd>
    </dl>
  </div>
</template>

<style lang="scss" scoped>
  .deck {
    margin-top: 50px;

    .desk {
      margin: 0 auto 30px;
    }

    .stats {
      font-size: rem(32px);
      margin: 0 auto;
      width: 730px;

      dt {
        float: left;
        clear: both;
        font-weight: bold;
      }

      dd {
        float: left;
        margin-left: rem(10px);

        &.has-full-houses {
          clear: both;
          float: none;
          display: block;
        }
      }

      .full-houses {
        margin: 0;
        list-style-type: lower-roman;
      }
    }
  }
</style>
