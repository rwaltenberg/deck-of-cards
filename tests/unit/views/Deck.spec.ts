import Vuex, { Store } from 'vuex'
import { createLocalVue, shallowMount } from '@vue/test-utils'

import Deck from '@/utils/deck'
import DeckDetails from '@/views/Deck.vue'

describe('Deck.vue', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)

  const mockDeckId = 'axj2sdfdc'
  let actions!: { fetchDeck: jest.Mock<any, any>}
  let store!: Store<any>
  let getters: { deck: () => Deck | null }

  beforeEach(() => {
    actions = { fetchDeck: jest.fn(async () => {}) }
    getters = {
      deck: () => (new Deck(['2H', '2D', '2C', '2S', '3H', '3D', '3C'], '2H', mockDeckId))
    }

    store = new Vuex.Store({ actions, getters })
  })

  it('does not try to fetch if already set on storage and route id matches', () => {
    const wrapper = shallowMount(
      DeckDetails,
      {
        store,
        localVue,
        mocks: {
          $route: {
            params: {
              id: mockDeckId
            }
          }
        }
      })

    expect(actions.fetchDeck).not.toHaveBeenCalled()
  })

  it('tries to fetch deck when there is none from store', () => {
    const emptyGetters = {
      deck: () => null
    }

    store = new Vuex.Store({ actions, getters: emptyGetters })

    const wrapper = shallowMount(
      DeckDetails,
      {
        store,
        localVue,
        mocks: {
          $route: {
            params: {
              id: mockDeckId
            }
          }
        }
      })

    expect(actions.fetchDeck).toHaveBeenCalled()
  })

  it('tries to fetch deck when deck id differs from route\'s deck id', () => {
    const wrapper = shallowMount(
      DeckDetails,
      {
        store,
        localVue,
        mocks: {
          $route: {
            params: {
              id: '3a2s1d5a3'
            }
          }
        }
      })

    expect(actions.fetchDeck).toHaveBeenCalled()
  })

  it('renders correctly', () => {
    const wrapper = shallowMount(
      DeckDetails,
      {
        store,
        localVue,
        mocks: {
          $route: {
            params: {
              id: mockDeckId
            }
          }
        }
      })

    expect(wrapper).toMatchSnapshot()
  })
})
