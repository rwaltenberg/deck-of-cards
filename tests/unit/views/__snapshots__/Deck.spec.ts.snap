// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck.vue renders correctly 1`] = `
<div class="deck">
  <desk-stub value="2,H,3,H,2,D,3,D,2,C,3,C,2,S"></desk-stub>
  <dl class="stats">
    <dt>High Card:</dt>
    <dd>2H</dd>
    <dt>Full House Combo:</dt>
    <dd class="has-full-houses">
      <ol class="full-houses">
        <li>2H, 2D, 2C, 3H, 3D</li>
        <li>2H, 2D, 2C, 3H, 3C</li>
        <li>2H, 2D, 2S, 3H, 3D</li>
        <li>2H, 2D, 2C, 3D, 3C</li>
        <li>2H, 2D, 2S, 3H, 3C</li>
        <li>2H, 2C, 2S, 3H, 3D</li>
        <li>2H, 2D, 2S, 3D, 3C</li>
        <li>2H, 2C, 2S, 3H, 3C</li>
        <li>2D, 2C, 2S, 3H, 3D</li>
        <li>2H, 2C, 2S, 3D, 3C</li>
        <li>2D, 2C, 2S, 3H, 3C</li>
        <li>2D, 2C, 2S, 3D, 3C</li>
        <li>2H, 2D, 3H, 3D, 3C</li>
        <li>2H, 2C, 3H, 3D, 3C</li>
        <li>2H, 2S, 3H, 3D, 3C</li>
        <li>2D, 2C, 3H, 3D, 3C</li>
        <li>2D, 2S, 3H, 3D, 3C</li>
        <li>2C, 2S, 3H, 3D, 3C</li>
      </ol>
    </dd>
  </dl>
</div>
`;
